### PR TITLE
Implementation of quartic & quintic splines with unit tests

### PR DESCRIPTION
**Pull request by AndrewHarris709**
Implements the quartic & quintic spline kernels, normalized in one, two, and three dimensions. Unit tests have been amended to test these new kernels.

Comparison of cubic, quartic, and quintic kernels (64x128 2D interpolation):
Cubic
image
Quartic
image
Quintic
image

As well, a new unit test has been created to test kernel normalization, using integration functions from the `scipy` module.

### CODE REVIEW

File: sarracen/tests/test_kernels.py [hunk 1]
```diff
@@ -1,22 +1,89 @@
...
+
+    def test_normalization(self):
+        for kernel in [CubicSplineKernel(), QuarticSplineKernel(), QuinticSplineKernel()]:
+            norm1 = quad(kernel.w, 0, kernel.radkernel, (1,))[0]
+            norm2 = dblquad(double_kernel, 0, kernel.radkernel, 0, kernel.radkernel, [kernel])[0]
+            norm3 = tplquad(triple_kernel, 0, kernel.radkernel, 0, kernel.radkernel, 0, kernel.radkernel, [kernel])[0]
+
+            assert approx(norm1) == 0.5  # a half of 1D space
+            assert approx(norm2) == 0.25  # a fourth of 2D space
+            assert approx(norm3) == 0.125 # an eight of 3D space
```
Comment (ttricco) [inline sarracen/tests/test_kernels.py:89]: This is great. Can you add some extra comments explaining why this equals 1/2, 1/4 and 1/8 in 1, 2, and 3 dimensions? Just that the integrals are over positive space, not the full volume of the kernel. At some point I'll look at this code again and have forgotten why.

File: sarracen/kernels/base_kernel.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    """A generic kernel used for data interpolation."""
+    def __init__(self, radkernel, cnormk1d, cnormk2d, cnormk3d, wfunc):
+        self._radkernel = radkernel
+        self._cnormk1d = cnormk1d
+        self._cnormk2d = cnormk2d
+        self._cnormk3d = cnormk3d
+        self._wfunc = wfunc
+
+    def w(self, q, dimension):
+        """
```
Comment (ttricco) [inline sarracen/kernels/base_kernel.py:11]: Actually a couple changes to suggest on how kernels are instantiated. Realistically, data is not going to change dimensions. So we don't need to call w() with the dimension every time. Let's change so the dimension is set in the constructor.

def __init__(self, radkernel, ndims, cnormk, wfunc)

def w(self, q)

Will need to refactor the constructors of each kernel to accept number of dimensions and choose the normalization constant based on that. Will also have to refactor the tests accordingly.

File: sarracen/kernels/cubic_spline.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+from sarracen.kernels import BaseKernel
+
+
+class CubicSplineKernel(BaseKernel):
+    """
+    An implementation of the Cubic Spline kernel, in 1, 2, and 3 dimensions.
+    """
+    def __init__(self):
+        super().__init__(2.0, 2./3., 10./(7.*np.pi), 1./np.pi, self._weight)
+
```
Comment (ttricco) [inline sarracen/kernels/cubic_spline.py:12]: Here (and in other kernels), this becomes

def __init__(self, ndims)

and the normalization constant is chosen based on the dimensionality and that value is passed to super().

**Review by ttricco — CHANGES_REQUESTED** [review]

File: sarracen/kernels/base_kernel.py (hunk 2 above)
Comment (AndrewHarris709) [reply sarracen/kernels/base_kernel.py:16]: Changed 👍 I also like how this removes the repetition of cnormk1d, cnormk2d, cnormk3d in the BaseKernel class

**Review by ttricco — APPROVED** [review]

**Pull request merged**